File: Lecture 05/Python Code/SRP/SRPFollowed.py
class Product:
    def __init__(self, name, price):
        self.name = name
        self.price = price

class ShoppingCart:
    def __init__(self):
        self.products = []

    def add_product(self, product):
        self.products.append(product)

    def calculate_total(self):
        return sum(p.price for p in self.products)

class ShoppingCartPrinter:
    def __init__(self, cart):
        self.cart = cart

    def print_invoice(self):
        print("Shopping Cart Invoice:")
        for product in self.cart.products:
            print(f"{product.name} - Rs {product.price}")
        print(f"Total: Rs {self.cart.calculate_total()}")

class ShoppingCartStorage:
    def __init__(self, cart):
        self.cart = cart

    def save_to_database(self):
        print("Saving shopping cart to database...")

if __name__ == "__main__":
    cart = ShoppingCart()
    cart.add_product(Product("Laptop", 50000))
    cart.add_product(Product("Mouse", 2000))

    printer = ShoppingCartPrinter(cart)
    printer.print_invoice()

    storage = ShoppingCartStorage(cart)
    storage.save_to_database()

File: Lecture 05/JavaScript Code/SRP/SRP_followed.js
// Product class representing any item of any ECommerce.
class Product {
    constructor(name, price) {
        this.name = name;
        this.price = price;
    }
}

// 1. ShoppingCart: Only responsible for cart management and business logic
class ShoppingCart {
    constructor() {
        this.products = [];
    }

    addProduct(product) {
        this.products.push(product);
    }

    getProducts() {
        return this.products;
    }

    // Business logic - calculating total belongs to cart
    calculateTotal() {
        let total = 0;
        for (const product of this.products) {
            total += product.price;
        }
        return total;
    }
}

// 2. InvoicePrinter: Only responsible for printing invoices
class InvoicePrinter {
    constructor(cart) {
        this.cart = cart;
    }

    printInvoice() {
        console.log("Shopping Cart Invoice:");
        for (const product of this.cart.getProducts()) {
            console.log(`${product.name} - Rs ${product.price}`);
        }
        console.log(`Total: Rs ${this.cart.calculateTotal()}`);
    }
}

// 3. DatabaseSaver: Only responsible for database operations
class DatabaseSaver {
    constructor(cart) {
        this.cart = cart;
    }

    saveToDatabase() {
        console.log("Saving shopping cart to database...");
        // If needed, access cart data: this.cart.getProducts()
    }
}


// Main execution
function main() {
    // Step 1: Create the cart
    const cart = new ShoppingCart();

    // Step 2: Add products to cart
    cart.addProduct(new Product("Laptop", 50000));
    cart.addProduct(new Product("Mouse", 2000));

    // Step 3: Create printer with cart reference and print
    const printer = new InvoicePrinter(cart);
    printer.printInvoice();

    // Step 4: Create database saver with cart reference and save
    const saver = new DatabaseSaver(cart);
    saver.saveToDatabase();
}

main();

